lib/core/src/Cardano/SharedWallet/DB/Sqlite.hs

```diff
@@ -0,0 +1,189 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- HLINT ignore "Redundant flip" -}
+
+-- |
+-- Copyright: © 2018-2021 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of the DBLayer which uses Persistent and SQLite.
+
+module Cardano.SharedWallet.DB.Sqlite
+    ( newDBLayer
+    , withDBLayer
+    , withDecoratedDBLayer
+    , DBDecorator (..)
+    , undecoratedDB
+    , defaultFilePath
+    ) where
+
+import Prelude
+
+import Cardano.DB.Sqlite
+    ( DBLog (..)
+    , MigrationError
+    , SqliteContext (..)
+    , newInMemorySqliteContext
+    , newSqliteContext
+    , withConnectionPool
+    )
+import Cardano.SharedWallet.DB
+    ( DBLayer (..) )
+import Cardano.SharedWallet.DB.Log
+    ( SharedWalletDbLog (..) )
+import Cardano.SharedWallet.DB.Sqlite.TH
+    ( migrateAll )
+import Control.Tracer
+    ( Tracer (..), contramap, traceWith )
+import Data.Function
+    ( (&) )
+import Database.Persist.Sqlite
+    ( SqlPersistT )
+import System.Directory
+    ( removeFile )
+import System.FilePath
+    ( (</>) )
+import UnliftIO.Exception
+    ( bracket, catch, throwIO )
+
+
+-- | Return the preferred @FilePath@ for the stake pool .sqlite file, given a
+-- parent directory.
+defaultFilePath
+    :: FilePath
+    -- ^ The directory in which the .sqlite file will be located.
+    -> FilePath
+defaultFilePath = (</> "shared-wallet.sqlite")
+
+-- | Runs an action with a connection to the SQLite database.
+--
+-- Database migrations are run to create tables if necessary.
+--
+-- If the given file path does not exist, it will be created by the sqlite
+-- library.
+withDBLayer
+    :: Tracer IO SharedWalletDbLog
+    -- ^ Logging object.
+    -> Maybe FilePath
+    -- ^ Database file location, or 'Nothing' for in-memory database.
+    -> (DBLayer IO k -> IO a)
+    -- ^ Action to run.
+    -> IO a
+withDBLayer = withDecoratedDBLayer undecoratedDB
+
+-- | A decorator for the database layer, useful for instrumenting or monitoring
+--   calls to database operations.
+newtype DBDecorator a k =
```

Comment:
DBDecorator is wrong and should be removed.

Reply:
ok, will do. Was trying to understand why it was added for Pool and bet it was very useful ;-)